
syntax = "proto3";

package Microsoft.Net.Remote.DataStream;

enum DataStreamOperationStatusCode
{
    DataStreamOperationStatusCodeUnknown = 0;
    DataStreamOperationStatusCodeSucceeded = 1;
    DataStreamOperationStatusCodeFailed = 2;
    DataStreamOperationStatusCodeCancelled = 3;
    DataStreamOperationStatusCodeTimedOut = 4;
}

message DataStreamOperationStatus
{
    DataStreamOperationStatusCode Code = 1;
    string Message = 2;
}

message DataStreamUploadData
{
    bytes Data = 1;
}

message DataStreamDownloadData
{
    bytes Data = 1;
    uint32 SequenceNumber = 2;
    DataStreamOperationStatus Status = 3;
}

message DataStreamUploadResult
{
    uint32 NumberOfDataBlocksReceived = 1;
    DataStreamOperationStatus Status = 2;
}

enum DataStreamType
{
    Fixed = 0;
    Continuous = 1;
}

enum DataStreamPattern
{
    Constant = 0;
}

message DataStreamFixedTypeProperties
{
    uint32 NumberOfDataBlocksToStream = 1;
}

message DataStreamContinuousTypeProperties
{

}

message DataStreamProperties
{
    DataStreamType Type = 1;
    DataStreamPattern Pattern = 2;
    oneof Properties
    {
        DataStreamFixedTypeProperties FixedTypeProperties = 3;
        DataStreamContinuousTypeProperties ContinuousTypeProperties = 4;
    }
}

message DataStreamDownloadRequest
{
    DataStreamProperties Properties = 1;
}
